### PR TITLE
Add integration tests - fixes #84

### LINKED ISSUE #84: Integration tests
We should run all the rules using ESLint against a huge amount of code to ensure it never crashes. We could include the source of some larger projects.

I usually manually run master against AVA to as an ensurance, but I forgot this time.

This could have prevented https://github.com/sindresorhus/eslint-plugin-unicorn/issues/81 and https://github.com/sindresorhus/eslint-plugin-unicorn/issues/82.

### PR DESCRIPTION
**Pull request by SamVerschueren**
This is still WIP, but feedback always welcome. Sorry for the `package.json` tabs vs spaces thing. Will fix it when everything is ready.

### CODE REVIEW

**Comment by SamVerschueren** [comment]
Tests are working now correctly. I defined the integration tests so that they are allowed to fail. I think it's more a notification for ourselves when new rules are introduced?

Any other packages that you guys like to see tested?

**Comment by jfmengels** [comment]
Nice! That works well!

A few things/ideas:

I understood the initial issue was about detecting crashes. Right now, the process fails because there are linting errors in the other projects, which is normal in this case because there are breaking changes, but normally that should not block the test. Do you know if ESLint fails returns a different exit code depending on whether there was a crash or whether there are linting errors? If not, I guess we could grep the output and check it manually.

That said, when not planning breaking changes, having the test fail when there are linting errors is great. Maybe we could have a flag that we add manually when we plan a major version that allows linting errors.

Ideally, we should test all rules with (almost) all possible configurations on, but that's hard because the projects we lint against do not enforce all the rules or with the alternative configurations. For the sake of testing crashes though, especially for new rules, there should probably be an additional test matrix like the following:
- For every project we test on
  - Test recommended config (**do not** allow linting errors, unless a major version is planned, but no crash)
  - Test config with all rules not in recommended or with alternative configurations (allow linting errors but no crashes). If there are multiple alternative configurations to test, we could create a matrix of configuration files to run the test with.

Alternatively, we could clone some of the projects once (`avajs/ava` --> `@SamVerschueren/ava-eslint-test-repo`), adapt them to follow the alternative configurations, and run the tests against those. Those would only need to be updated at every major version.

Anyway, this is a great start and my comments could perhaps be the basis of a new issue/PR rather than integrate this all into this PR. Great work @SamVerschueren!

**Comment by SamVerschueren** [comment]
> I understood the initial issue was about detecting crashes. Right now, the process fails because there are linting errors in the other projects

A good catch! You are right, it shouldn't fail when the code does not passes the linter, it should fail when a rule forgets something and it crashes.

> Do you know if ESLint fails returns a different exit code depending on whether there was a crash or whether there are linting errors? If not, I guess we could grep the output and check it manually.

We'll have to check it manually, I couldn't find a difference.

> That said, when not planning breaking changes, having the test fail when there are linting errors is great. Maybe we could have a flag that we add manually when we plan a major version that allows linting errors.

Think that might be hard to maintain. Because after a breaking change release, we'll have to turn off the flag again. If we forget that, it will silently pass. If you have any ideas on that, feel free to elaborate :).

> Ideally, we should test all rules with (almost) all possible configurations on

You're right, good suggestion. I'm going to think about how we can do that in an easy manner. Think I already have some ideas without forking or whatsoever. But maybe let's focus on the initial use case and improve further in later PRs like you suggest.

---------------------------
So the next thing that I will do is make sure the integration tests just passes when no actual error occurred (and remove the `allow_failures` flag in `.travis.yml`.

Any additional feedback? Extra repositories that we can lint?

File: test/integration/test.js [hunk 1]
```diff
@@ -0,0 +1,68 @@
...
+'use strict';
+const path = require('path');
+const Listr = require('listr');
+const tempy = require('tempy');
+const execa = require('execa');
+const del = require('del');
+
+const packages = new Map([
+	['got', 'https://github.com/sindresorhus/got'],
+	['ava', 'https://github.com/avajs/ava']
```
Comment (sindresorhus) [inline test/integration/test.js]: Some more:

- https://github.com/sindresorhus/pageres
- https://github.com/sindresorhus/np
- https://github.com/sindresorhus/ora
- https://github.com/sindresorhus/p-map
- https://github.com/sindresorhus/detect-indent
- https://github.com/sindresorhus/os-locale
- https://github.com/sindresorhus/execa
- https://github.com/sindresorhus/pify
- https://github.com/sindresorhus/boxen
- https://github.com/sindresorhus/make-dir


@kevva Can you list some of your larger repos?

@SamVerschueren You should add `listr` and some of your other repos.

File: test/integration/test.js [hunk 2]
```diff
@@ -0,0 +1,68 @@
...
+		}
+	], {
+		exitOnError: false
+	});
+};
+
+const list = new Listr([
+	{
+		title: 'Setup',
+		task: () => execa('npm', ['install', '../../../', 'eslint'], {cwd})
```
Comment (sindresorhus) [inline test/integration/test.js]: `../../../` => `../../..`

File: test/integration/test.js [hunk 3]
```diff
@@ -0,0 +1,68 @@
...
+	},
+	{
+		title: 'Integration tests',
+		task: () => {
+			const tests = new Listr({concurrent: true});
+
+			for (const [name, url] of packages.entries()) {
```
Comment (sindresorhus) [inline test/integration/test.js]: `.entries()` is moot here. Just do `for (const [name, url] of packages) {`

File: test/integration/test.js [hunk 4]
```diff
@@ -0,0 +1,68 @@
...
+	}
+]);
+
+list.run()
+	.catch(err => {
+		for (const error of err.errors) {
+			console.error(error.message);
+		}
+
+		process.exit(-1);
```
Comment (sindresorhus) [inline test/integration/test.js]: Why `-1`?

File: .travis.yml [hunk 5]
```diff
@@ -3,6 +3,15 @@ node_js:
...
+  include:
+    - node_js: 8
+      env: INTEGRATION=true
+  allow_failures:
+    - node_js: 8
+      env: INTEGRATION=true
 before_install:
   - npm i -g npm
+script:
+  - if [ "$INTEGRATION" = true ]; then npm run integration else npm test; fi
```
Comment (sindresorhus) [inline .travis.yml]: Can you quote the commands? It's a bit hard to read without quotes or syntax highlighting.

File: test/integration/test.js (hunk 4 above)
Comment (SamVerschueren) [inline test/integration/test.js]: Should've been `1` :).

File: .travis.yml (hunk 5 above)
Comment (SamVerschueren) [inline .travis.yml]: What do you mean with `quote the commands`? I can't do this:

```
- if [ "$INTEGRATION" = true ]; then "npm run integration" else "npm test"; fi
```

I can do this if you like it more

```
- if [ "$INTEGRATION" = true ]; then npm run integration; fi
- if [ -z "$INTEGRATION ]; then npm test; fi
```

Not sure about the `-z`, have to verify. But you'll get the idea :).

Comment (sindresorhus) [inline .travis.yml]: > - if [ "$INTEGRATION" = true ]; then "npm run integration" else "npm test"; fi

Oh, I assumed that would work.

How about:

```
 [ "$INTEGRATION" = true ] && npm run integration || npm test
```

File: test/integration/test.js [hunk 6]
```diff
@@ -0,0 +1,92 @@
...
+	const dest = tempy.directory();
+
+	return new Listr([
+		{
+			title: 'Cloning',
+			task: () => execa('git', ['clone', packages.get(name), '--single-branch', dest])
+		},
+		{
+			title: 'Running tests',
+			task: () => execa('./node_modules/.bin/eslint', ['--config', path.join(cwd, 'index.js'), dest], {cwd})
```
Comment (sindresorhus) [inline test/integration/test.js]: You could use the `localDir` option here.

File: test/integration/test.js [hunk 7]
```diff
@@ -0,0 +1,92 @@
...
+				.catch(err => {
+					if (!/✖ [0-9]+ problems? \([0-9]+ errors?, [0-9]+ warnings?\)/.test(err.message)) {
```
Comment (sindresorhus) [inline test/integration/test.js]: `[0-9]` => `\d`

File: package.json [hunk 8]
```diff
@@ -1,70 +1,81 @@
...
+    "xo": "*"
+  },
+  "peerDependencies": {
+    "eslint": ">=4"
+  },
+  "ava": {
+    "files": [
+      "test/*.js"
+    ]
+  }
```
Comment (sindresorhus) [inline package.json]: You changed the indentation of package.json

Comment (SamVerschueren) [inline package.json]: I don't use npm@5 yet :p

File: .travis.yml (hunk 5 above)
Comment (SamVerschueren) [inline .travis.yml]: Doesn't seem to be valid yaml https://travis-ci.org/sindresorhus/eslint-plugin-unicorn/builds/256575546

File: test/integration/test.js (hunk 7 above)
Comment (SamVerschueren) [inline test/integration/test.js]: Fixed. If you have a better idea on how to find out if the error are ESLint warnings or something else, feel free to let me know.

File: test/integration/test.js (hunk 6 above)
Comment (SamVerschueren) [inline test/integration/test.js]: Good catch!

File: .travis.yml (hunk 5 above)
Comment (sindresorhus) [inline .travis.yml]: Ah, you have to quote it:

```
  - "[ \"$INTEGRATION\" = true ] && npm run integration || npm test"
```

Another reason I hate YAML. You never know when you have to quote or not.

Comment (SamVerschueren) [inline .travis.yml]: Ugh! That seems to work, thanks!

File: test/integration/test.js (hunk 1 above)
Comment (kevva) [inline test/integration/test.js]: I don't have that many large repos anymore, but here are some more suggestions:

* https://github.com/kevva/to-ico
* https://github.com/kevva/download
* https://github.com/kevva/brightness
* https://github.com/kevva/decompress
* https://github.com/kevva/npm-conf
* https://github.com/imagemin/imagemin
* https://github.com/chalk/wrap-ansi

Comment (sindresorhus) [inline test/integration/test.js]: And https://github.com/chalk/chalk https://github.com/qix-/color-convert

Comment (SamVerschueren) [inline test/integration/test.js]: Added! Thanks guys!

File: test/integration/test.js [hunk 9]
```diff
@@ -0,0 +1,102 @@
...
+	['bragg-router', 'http://github.com/SamVerschueren/bragg-router'],
+	['dev-time', 'http://github.com/SamVerschueren/dev-time'],
+	['decode-uri-component', 'https://github.com/SamVerschueren/decode-uri-component'],
+	['to-ico', 'https://github.com/kevva/to-ico'],
+	['download', 'https://github.com/kevva/download'],
+	['brightness', 'https://github.com/kevva/brightness'],
+	['decompress', 'https://github.com/kevva/decompress'],
+	['npm-conf', 'https://github.com/kevva/npm-conf'],
+	['imagemin', 'https://github.com/imagemin/imagemin'],
+	['ctolor-conver', 'https://github.com/qix-/color-convert']
```
Comment (sindresorhus) [inline test/integration/test.js]: typo

Comment (SamVerschueren) [inline test/integration/test.js]: Doh, thanks :)

**Comment by SamVerschueren** [comment]
Anything else I can do for the integration tests?

**Comment by sindresorhus** [comment]
🙌 It's great!

**Comment by SamVerschueren** [comment]
Thanks for the feedback guys! :)

**Pull request merged**